File: dart/lib/src/measurement.dart
/// Represents a single measurement with statement and duration
class Measurement {
  /// The statement or operation name
  final String statement;
  
  /// The duration of the operation
  final Duration duration;

  const Measurement({
    required this.statement,
    required this.duration,
  });

  @override
  String toString() => 'Measurement(statement: $statement, duration: $duration)';

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Measurement &&
          runtimeType == other.runtimeType &&
          statement == other.statement &&
          duration == other.duration;

  @override
  int get hashCode => statement.hashCode ^ duration.hashCode;
}

/// Represents a group of similar measurements
class GroupedMeasurement {
  /// The grouped operation name
  final String name;
  
  /// Number of operations in this group
  final int count;
  
  /// Total time of all operations
  final Duration totalTime;
  
  /// Average time per operation
  final Duration avgTime;
  
  /// Minimum time in the group
  final Duration minTime;
  
  /// Maximum time in the group
  final Duration maxTime;

  const GroupedMeasurement({
    required this.name,
    required this.count,
    required this.totalTime,
    required this.avgTime,
    required this.minTime,
    required this.maxTime,
  });

  @override
  String toString() => 'GroupedMeasurement(name: $name, count: $count, avgTime: $avgTime)';
}

File: swift/Sources/QuickTrace/Measurement.swift
import Foundation

/// Represents a single measurement in the trace
public struct Measurement {
    /// The name/description of the operation
    public let statement: String
    
    /// The duration of the operation
    public let duration: TimeInterval
    
    /// Creates a new measurement
    /// - Parameters:
    ///   - statement: The name/description of the operation
    ///   - duration: The duration of the operation in seconds
    public init(statement: String, duration: TimeInterval) {
        self.statement = statement
        self.duration = duration
    }
}

extension Measurement: Equatable {
    public static func == (lhs: Measurement, rhs: Measurement) -> Bool {
        return lhs.statement == rhs.statement && abs(lhs.duration - rhs.duration) < 0.000001
    }
}

extension Measurement: CustomStringConvertible {
    public var description: String {
        return "\(statement): \(String(format: "%.3f", duration))s"
    }
}

/// Represents a group of similar measurements
public struct GroupedMeasurement {
    /// The name of the group
    public let name: String
    
    /// Number of measurements in this group
    public let count: Int
    
    /// Total time of all measurements
    public let totalTime: TimeInterval
    
    /// Average time of measurements in this group
    public let avgTime: TimeInterval
    
    /// Minimum time in this group
    public let minTime: TimeInterval
    
    /// Maximum time in this group
    public let maxTime: TimeInterval
    
    /// Creates a new grouped measurement
    public init(name: String, count: Int, totalTime: TimeInterval, avgTime: TimeInterval, minTime: TimeInterval, maxTime: TimeInterval) {
        self.name = name
        self.count = count
        self.totalTime = totalTime
        self.avgTime = avgTime
        self.minTime = minTime
        self.maxTime = maxTime
    }
}

extension GroupedMeasurement: CustomStringConvertible {
    public var description: String {
        return "\(name) (×\(count)): avg \(String(format: "%.3f", avgTime))s, range \(String(format: "%.3f", minTime))s-\(String(format: "%.3f", maxTime))s"
    }
}

